src/plugin/adapter-registry/c-adapter.ts

```diff
@@ -0,0 +1,52 @@
+/********************************************************************************
+ * Copyright (C) 2023 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import * as vscode from 'vscode';
+import * as manifest from '../manifest';
+import { VariableTracker } from './adapter-capabilities';
+import { outputChannelLogger } from '../logger';
+import { AdapterRegistry } from './adapter-registry';
+import { CTracker } from './c-tracker';
+
+export class CAdapter {
+    protected disposable: vscode.Disposable | undefined;
+
+    constructor(protected registry: AdapterRegistry) {
+    }
+
+    activate(context: vscode.ExtensionContext): void {
+        vscode.workspace.onDidChangeConfiguration(e => {
+            if (e.affectsConfiguration(`${manifest.PACKAGE_NAME}.${manifest.CONFIG_DEBUG_TYPES}`)) {
+                this.register(context);
+            }
+        });
+
+        this.register(context);
+    }
+
+    protected register(context: vscode.ExtensionContext): void {
+        if (this.disposable) {
+            this.disposable.dispose();
+        }
+
+        const debugTypes = vscode.workspace.getConfiguration(manifest.PACKAGE_NAME).get<string[]>(manifest.CONFIG_DEBUG_TYPES) || manifest.DEFAULT_DEBUG_TYPES;
+        const tracker = new VariableTracker(CTracker, outputChannelLogger, ...debugTypes);
+
+        const disposable = this.registry.registerAdapter(tracker, ...debugTypes);
```

Comment:
I guess this is likely where the lifecycle issue comes up - we replace the `VariableTracker` on reregistration, so the new one wasn't created to track variables for the existing session. Raises a couple of questions:

 - Do both of these need to know the new debug types? I think it was likely a little overzealous of me to include it in both places: perhaps the VariableTracker could just assume that whotever's passing it a session knows what it's doing? That's what the registration with the registry is for: matching sessions to trackers that can handle them.
 - How do we want to handle preference changes when a session is running? The current logic basically eliminates variable tracking for everybody if there's a preference change in a running session. We could defer handling it till running sessions were terminated. We could figure out if there's a more graceful way to handle it that acknowledges the new value without changing things for existing sessions.

Reply:
> I guess this is likely where the lifecycle issue comes up - we replace the VariableTracker on reregistration, so the new one wasn't created to track variables for the existing session.

Yes, this works as expected when changed outside of a debug session.

I think this is fixable by parsing out the changes and only deleting/re-registering the changed debuggers. However I'm not sure we want to make such fundamental changes during a running session. I think it should be enough to state `debug restart required` next to the setting.

WDYT?

Reply:
I think that's fair for now 👍

Reply:
Added